File: leetcode/src/combinationsumII.cpp
/*
 * given a set of candidate numbers (C) and a target number (T), find all unique combinations in C where candidate number sum to T.
 * ps, 
 * 1.the same number in candidates can be chosen from C with only once
 * 2.all numbers are positive integers with repeated value in possible
 * 3.all elements in a combination must be non-decending order without duplicate combinations.
 * */
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution{

public:
    vector<vector<int> > combinationSumII(vector<int>& num, int target){
        sort(num.begin(), num.end());  //sort in ascending order
        vector<vector<int> > res;
        vector<int> vec;
        plusSum(res, num, vec, target);
        return res;
    }

private:
    void plusSum(vector<vector<int> >& res, const vector<int>& candidates, vector<int>& vec, int target){
        if(target == 0){
            res.push_back(index2Values(candidates, vec));
            return;
        }

        const int n = candidates.size();
        int start = vec.empty() ? 0 : vec[vec.size() - 1] + 1;
        for(int i = start; i < n; ++i){
            if(target < candidates[i])        break;

            if(i > start &&  candidates[i] == candidates[i-1])    continue;  // skip duplicat combinations

            vec.push_back(i); // vec<> stores the index in candidates<>
            plusSum(res, candidates, vec, target - candidates[i]);
            vec.pop_back();
        }
        return;
    }

    vector<int> index2Values(const vector<int>& candidates, vector<int>& indexes){
        vector<int> values;
        int n = indexes.size();
        for(int i = 0; i < n; ++i){
            values.push_back(candidates[indexes[i]]);
        }
        return values;
    }
};

File: leetcode/java/combinationsumII.java
/*
 * given a collection of candidate numbers (C) and a target number (T), find all unique combination in C
 * where the candidate numbers sums to T.
 *
 * Note:
 * 0. each number in C can be used only once in one combination.
 * 1. all numbers (including target) will be positive integers
 * 2. elements in a combination result must be in non-descending order
 * 3. the result set cannot have duplicate combinations
 * by the way, C can have duplciate numbers
 *
 * test data: 
 * {1,1,2,2,2,3}, target = 6
 * result: {1,1,2,2}, {1,2,3}, {2,2,2}
 * */
import java.io.*;
import java.util.*;

public class combinationsumII{
    private Stack<Integer> stk = new Stack<Integer>();  // push index of num[]

    private void putStack(int[] num, ArrayList<ArrayList<Integer>> combinations){
        ArrayList<Integer> arr = new ArrayList<Integer>();
        for(int i : stk){
            arr.add(num[i]); // save number instead of index in result
        }
        combinations.add(arr);
        return;
    }

    private void plusSum(int[] num, int target, ArrayList<ArrayList<Integer>> result){
        if(target == 0){
            putStack(num, result);
            return;
        }

        int start = stk.isEmpty() ? 0 : stk.peek() + 1;
        /// num[i] is the first number which can be used for consistent args (stk, target)
        for(int i = start; i < num.length; ++i){
            if(target < num[i])    break;
            if(i > start && num[i] == num[i-1])    continue; // skip duplicate combination!
            stk.push(i);
            plusSum(num, target - num[i], result);
            stk.pop();
        }
        return;
    }

    public ArrayList<ArrayList<Integer>> combinationSumII(int[] num, int target){
        Arrays.sort(num);    //sort in ascending order
        ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>>();
        stk.clear();
        plusSum(num, target, result);
        return result;
    }

    /* unit test is in ../java_unittest/combinatiomsumII_junit/  */

    public void display(ArrayList<ArrayList<Integer>> combinations){
        System.out.println("combinations:");
        for(ArrayList<Integer> arr : combinations){
            for(Integer i : arr){
                System.out.print(i + " ");
            }
            System.out.println();
        }
    }

    public void test_01(){
        Scanner scan = new Scanner(System.in);
        while(true){
            System.out.println("please input integer candidates:");
            String str = scan.nextLine().trim();
            if(str.isEmpty())    break;
            StringTokenizer t = new StringTokenizer(str, " ,");
            int[] nums = new int[t.countTokens()];
            int i = 0;
            while(t.hasMoreTokens()){
                nums[i++] = Integer.parseInt(t.nextToken().toString());
            }

            System.out.println("please input integer target:");
            str = scan.nextLine().trim();
            if(str.isEmpty())    break;
            int x = Integer.parseInt(str);

            display(combinationSumII(nums, x));
        }
        return;
    }
}
